cabal-install/Distribution/Client/Dependency.hs

```diff
@@ -346,7 +348,7 @@ dontUpgradeNonUpgradeablePackages params =
   where
     extraConstraints =
       [ LabeledPackageConstraint
-        (PackageConstraintInstalled pkgname)
+        (PackageConstraint (unqualified pkgname) PackagePropertyInstalled)
```

Comment:
On second thoughts, having looked at how `Qualifier` is used in the solver, I think my last suggestion was incorrect. So, if I understand you, I should modify `PackageConstraint` to support a wildcard or 'any' qualifier as an extra alternative, but leave the `Qualifier` and `QPN` datatypes as they are?